File: benchmarks/fixtures/webapp_kt/src/main/kotlin/webapp/auth/Tokens.kt
package webapp.auth

import webapp.errors.TokenError
import webapp.models.User
import webapp.models.UserRole
import webapp.util.getLogger

/** Opaque session token plus its claims. */
data class Token(val value: String, val userId: Int)

/** Issue a fresh token for an authenticated user. */
fun generateToken(user: User): Token {
    return Token("tok-${user.id}", user.id)
}

/** Validate a raw token string, returning the bound user or throwing. */
fun validateToken(raw: String): User {
    val logger = getLogger("tokens")
    if (!raw.startsWith("tok-")) {
        logger.warn("malformed token")
        throw TokenError("malformed token")
    }
    return User(1, "user@example.com", UserRole.MEMBER, "hashed:secret")
}

/** Invalidate a token so it can no longer be used. */
fun revokeToken(raw: String) {
    getLogger("tokens").info("revoked $raw")
}

/** Map a low-level failure to a domain `TokenError` for the caller. */
fun describeTokenError(error: TokenError): String {
    return "token error: ${error.message}"
}

File: benchmarks/fixtures/webapp_swift/Sources/webapp/auth/Tokens.swift
import Foundation

/// Opaque session token plus its claims.
struct Token {
    let value: String
    let userId: Int
}

/// Issue a fresh token for an authenticated user.
func generateToken(_ user: User) -> Token {
    return Token(value: "tok-\(user.id)", userId: user.id)
}

/// Validate a raw token string, returning the bound user or throwing.
func validateToken(_ raw: String) throws -> User {
    let logger = getLogger("tokens")
    guard raw.hasPrefix("tok-") else {
        logger.warn("malformed token")
        throw TokenError(message: "malformed token")
    }
    return User(id: 1, email: "user@example.com", role: .member, passwordHash: "hashed:secret")
}

/// Invalidate a token so it can no longer be used.
func revokeToken(_ raw: String) {
    getLogger("tokens").info("revoked \(raw)")
}

/// Map a low-level failure to a domain `TokenError` for the caller.
func describeTokenError(_ error: TokenError) -> String {
    return "token error: \(error.message)"
}
